R/parse.R

```diff
@@ -51,18 +51,24 @@ parse_delim <- function(x, delim, quote = "\"", ...) {
   )
 }
 
-parse_name_equals_value <- function(x) {
-  loc <- regexpr("=", x, fixed = TRUE)
-  pieces <- regmatches(x, loc, invert = TRUE)
+parse_name_equals_value <- function(x, safe = "&", quote = "\"") {
+  if (length(x) == 0) {
+    return(NULL)
+  }
+  # Note: quotes are removed in the parse_delim function
+  # re-introduce quotes to create a safe pattern to split on
+  safe_pattern <- paste0(quote, safe, quote)
+  pieces <- strsplit(sub("=", safe_pattern, x, fixed = T), safe_pattern, fixed = TRUE)
+  pieces_matrix <- do.call(rbind, pieces)
+
+  if (ncol(pieces_matrix) == 1) {
+    pieces_matrix <- cbind(pieces_matrix, rep("", nrow(pieces_matrix)))
+  }
 
   # If only one piece, assume it's a field name with empty value
-  expand <- function(x) if (length(x) == 1) c(x, "") else x
-  pieces <- map(pieces, expand)
-
-  val <- trimws(map_chr(pieces, "[[", 2))
-  name <- trimws(map_chr(pieces, "[[", 1))
-
-  set_names(as.list(val), name)
+  found <- pieces_matrix[, 1] == pieces_matrix[, 2]
```

Comment:
This happens because of recycling in `rbind()`?

Reply:
Yes for example:

``` r
ll <- list(
  c("var1", "foo"),
  c("var2"),
  c("var3", "bar")
)

do.call(rbind, ll)
#>      [,1]   [,2]  
#> [1,] "var1" "foo" 
#> [2,] "var2" "var2"
#> [3,] "var3" "bar"
```

<sup>Created on 2024-01-29 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>